fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/InterestPeriod.java

```diff
@@ -0,0 +1,110 @@
+        this.repaymentPeriod = repaymentPeriod;
+        this.fromDate = fromDate;
+        this.dueDate = dueDate;
+        this.rateFactor = rateFactor;
+        this.disbursementAmount = disbursementAmount;
+        this.balanceCorrectionAmount = balanceCorrectionAmount;
+        this.outstandingLoanBalance = outstandingLoanBalance;
+    }
+
+    public InterestPeriod(RepaymentPeriod repaymentPeriod, InterestPeriod interestPeriod) {
+        this(repaymentPeriod, interestPeriod.getFromDate(), interestPeriod.getDueDate(), interestPeriod.getRateFactor(),
+                interestPeriod.getDisbursementAmount(), interestPeriod.getBalanceCorrectionAmount(),
+                interestPeriod.getOutstandingLoanBalance());
+    }
+
+    @Override
+    public int compareTo(@NotNull InterestPeriod o) {
+        return dueDate.compareTo(o.dueDate);
+    }
+
+    public void addBalanceCorrectionAmount(final Money balanceCorrectionAmount) {
+        this.balanceCorrectionAmount = MathUtil.plus(this.balanceCorrectionAmount, balanceCorrectionAmount);
+    }
+
+    public void addDisbursementAmount(final Money disbursementAmount) {
+        this.disbursementAmount = MathUtil.plus(this.disbursementAmount, disbursementAmount);
+    }
+
+    public Money getCalculatedDueInterest() {
+        return getOutstandingLoanBalance().multipliedBy(getRateFactor());
+    }
+
+    public void updateOutstandingLoanBalance() {
+        if (isFirstInterestPeriod()) {
+            Optional<RepaymentPeriod> previousRepaymentPeriod = getRepaymentPeriod().getPrevious();
+            if (previousRepaymentPeriod.isPresent()) {
+                InterestPeriod previousInterestPeriod = previousRepaymentPeriod.get().getInterestPeriods()
+                        .get(previousRepaymentPeriod.get().getInterestPeriods().size() - 1);
+                this.outstandingLoanBalance = previousInterestPeriod.getOutstandingLoanBalance()//
+                        .plus(previousInterestPeriod.getDisbursementAmount())//
+                        .plus(previousInterestPeriod.getBalanceCorrectionAmount())//
+                        .minus(previousRepaymentPeriod.get().getDuePrincipal())//
+                        .plus(previousRepaymentPeriod.get().getPaidPrincipal());//
+            }
+        } else {
+            int index = getRepaymentPeriod().getInterestPeriods().indexOf(this);
+            InterestPeriod previousInterestPeriod = getRepaymentPeriod().getInterestPeriods().get(index - 1);
+            this.outstandingLoanBalance = previousInterestPeriod.getOutstandingLoanBalance() //
+                    .plus(previousInterestPeriod.getBalanceCorrectionAmount()) //
+                    .plus(previousInterestPeriod.getDisbursementAmount()); //
+        }
+    }
+
+    private boolean isFirstInterestPeriod() {
+        return getRepaymentPeriod().getInterestPeriods().get(0).equals(this);
+    }
+
+    @Override
+    public String toString() {
```

Comment:
Why not using a Lombok toString?

Reply:
Originally it was part of a debug purpose and now I will remove.